azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -128,40 +122,64 @@ def update_host_plugin_headers(wire_client):
         # Fetching the goal state updates the HostGAPlugin so simply trigger the request
         GoalState._fetch_goal_state(wire_client)
 
-    def update(self, force_update=False):
+    def update(self):
         """
         Updates the current GoalState instance fetching values from the WireServer/HostGAPlugin as needed
         """
-        timestamp = create_timestamp()
-        xml_text, xml_doc, incarnation = GoalState._fetch_goal_state(self._wire_client)
-
-        if force_update or self._incarnation != incarnation:
-            # If we are fetching a new goal state
-            self._history = GoalStateHistory(timestamp, incarnation)
-            self._initialize_basic_properties(xml_doc)
-            self._fetch_extended_goal_state(xml_text, xml_doc, force_vm_settings_update=force_update)
-        else:
-            # else ensure the extensions are using the latest vm_settings
-            timestamp = create_timestamp()
-            vm_settings, vm_settings_updated = self._fetch_vm_settings(force_update=force_update)
+        #
+        # Fetch the goal state from both the HGAP and the WireServer
+        #
+        timestamp = timeutil.create_timestamp()
+
+        incarnation, xml_text, xml_doc = GoalState._fetch_goal_state(self._wire_client)
+        goal_state_updated = incarnation != self._incarnation
+        if goal_state_updated:
+            logger.info('Fetched new goal state from the WireServer [incarnation {0}]', incarnation)
+
+        vm_settings, vm_settings_updated = GoalState._fetch_vm_settings(self._wire_client)
+        if vm_settings_updated:
+            logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
+        # Ignore the vmSettings if their source is Fabric (processing a Fabric goal state may require the tenant certificate and the vmSettings don't include it.)
+        if vm_settings is not None and vm_settings.source == GoalStateSource.Fabric:
             if vm_settings_updated:
-                logger.info("Fetched new vmSettings [HostGAPlugin correlation ID: {0} eTag: {1} source: {2}]", vm_settings.hostga_plugin_correlation_id, vm_settings.etag, vm_settings.source)
-                self._history = GoalStateHistory(timestamp, vm_settings.etag)
-                self._extensions_goal_state = vm_settings
-                self._history.save_vm_settings(vm_settings.get_redacted_text())
+                logger.info("The vmSettings originated via Fabric; will ignore them.")
+            vm_settings, vm_settings_updated = None, False
+
+        # If neither goal state has changed we are done with the update
+        if not goal_state_updated and not vm_settings_updated:
+            return
+
+        # Start a new history subdirectory and capture the updated goal state
```

Comment:
Note that now the suffix for the history directories is _incarnation-etag

Reply:
wont this get confusing?

Reply:
confusing in what sense? I posted a sample fragment below.

for reference
`
./2022-03-31T22_26_21.054052_4:
    -rwxrwxrwx 1 nam nam 2800 Mar 31 22:26 ExtensionsConfig.xml
    -rwxrwxrwx 1 nam nam 2099 Mar 31 22:26 GoalState.xml
    -rwxrwxrwx 1 nam nam 1441 Mar 31 22:26 HostingEnvironmentConfig.xml
    -rwxrwxrwx 1 nam nam 7406 Mar 31 22:26 Microsoft.Azure.Extensions.CustomScript.manifest.xml
    -rwxrwxrwx 1 nam nam  899 Mar 31 22:26 SharedConfig.xml
    -rwxrwxrwx 1 nam nam 1020 Mar 31 22:26 waagent_status.json

./2022-03-31T22_27_01.331737_5:
    -rwxrwxrwx 1 nam nam  3187 Mar 31 15:27 ExtensionsConfig.xml
    -rwxrwxrwx 1 nam nam  2099 Mar 31 15:27 GoalState.xml
    -rwxrwxrwx 1 nam nam  1721 Mar 31 15:27 HostingEnvironmentConfig.xml
    -rwxrwxrwx 1 nam nam 13020 Mar 31 15:27 Microsoft.OSTCExtensions.VMAccessForLinux.manifest.xml
    -rwxrwxrwx 1 nam nam   899 Mar 31 15:27 SharedConfig.xml
    -rwxrwxrwx 1 nam nam  1632 Mar 31 15:27 waagent_status.json

./2022-03-31T22_27_47.722244_5-11145155257095352179:
    -rwxrwxrwx 1 nam nam 13020 Mar 31 15:27 Microsoft.OSTCExtensions.VMAccessForLinux.manifest.xml
    -rwxrwxrwx 1 nam nam  2299 Mar 31 15:27 VmSettings.json
    -rwxrwxrwx 1 nam nam  1018 Mar 31 15:28 waagent_status.json

./2022-03-31T22_28_58.078953_5-1812135654488544646:
    -rwxrwxrwx 1 nam nam 1301 Mar 31 15:28 Microsoft.CPlat.Core.RunCommandHandlerLinux.manifest.xml
    -rwxrwxrwx 1 nam nam 2527 Mar 31 15:28 VmSettings.json
    -rwxrwxrwx 1 nam nam 1845 Mar 31 15:29 waagent_status.json`

Reply:
I primarily meant like originally we were only saving the etag/incarnation of the one we executed vs both. I just figured keeping both might get confusing but I'm not averse to it. We can deploy it as is and if it does get confusing to debug we can always revisit this

Reply:
when there is a fast track goal state that requires the certificates of a previous fabric goal state having the incarnation/etag combo is very handy. In the above example if there was an issue in etag 1812135654488544646, a glance to the history folder can point out very easily when there were changes to the wireserver goal state (one just needs to glance at changes in the incarnation)

but, as you point out, we can always iterate on this depending on how debugging of actual issues goes